tests/phpunit/FileSystem/FileSystemTestCase.php

```diff
@@ -0,0 +1,126 @@
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the box project.
+ *
+ * (c) Kevin Herrera <kevin@herrera.io>
+ *     Théo Fidry <theo.fidry@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Infection\Tests\FileSystem;
+
+use function array_map;
+use function array_values;
+use function chdir;
+use const DIRECTORY_SEPARATOR;
+use function getcwd;
+use function Infection\Tests\make_tmp_dir;
+use function Infection\Tests\normalizePath;
+use function natcasesort;
+use PHPUnit\Framework\TestCase;
+use function realpath;
+use function str_replace;
+use Symfony\Component\Filesystem\Filesystem;
+use function sys_get_temp_dir;
+
+/**
+ * @private
+ */
+abstract class FileSystemTestCase extends TestCase
+{
+    /** @var string */
+    protected $cwd;
+
+    /** @var string */
+    protected $tmp;
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        // Cleans up whatever was there before. Indeed upon failure PHPUnit fails to trigger the `tearDown()` method
+        // and as a result some temporary files may still remain.
+        (new Filesystem())->remove(
+            normalizePath(realpath(sys_get_temp_dir()) . '/infection-test')
+        );
+
+        $this->cwd = getcwd();
+        $this->tmp = make_tmp_dir('infection-test', self::class);
+
+        chdir($this->tmp);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function tearDown(): void
```

Comment:
I may be wrong here, but i think `tearDown` is not called after a failed test. Meaning we also need the clean up in a `tearDownAfterClass`

Reply:
I don't know if `tearDownAfterClass()` is called after a failing test either... But `teardown()` is definitely not. It should be done here still though in case you require a clean environment in-between each test method. Also the workspace used is in a deterministic base tmp dir hence, since we remove that base dir at the beginning, we always limit the number of unclean artefact

Reply:
Also note that FS is never transactional, so there is no safe way to guarantee a perfect environment with no artefact leftover. That's why the strategy here is more to limit where those artefacts can be and clean them up whenever we can (see the `setUp()`)

Reply:
It should indeed be done here, i'm suggesting to duplicate it in a `tearDownAfterClass`.

Reply:
Done